webpack/webpack-workspace-resolve-plugin/src/WorkspaceLayoutCache.ts

```diff
@@ -0,0 +1,198 @@
+ * @beta
+ */
+export interface IResolverCacheFile {
+  /**
+   * The ordered list of all contexts in the cache
+   */
+  contexts: ISerializedResolveContext[];
+}
+
+/**
+ * A context for resolving dependencies in a workspace.
+ * @beta
+ */
+export interface IResolveContext {
+  /**
+   * The absolute path to the root folder of this context
+   */
+  descriptionFileRoot: string;
+  /**
+   * Find the context that corresponds to a module specifier, when requested in the current context.
+   * @param request - The module specifier to resolve
+   */
+  findDependency(request: string): IPrefixMatch<IResolveContext> | undefined;
+}
+
+/**
+ * Options for creating a `WorkspaceLayoutCache`.
+ * @beta
+ */
+export interface IWorkspaceLayoutCacheOptions {
+  /**
+   * The root folder of the workspace. All paths in the cache file are assumed to be relative to this folder.
+   */
+  workspaceRoot: string;
+  /**
+   * The parsed cache data. File reading is left as an exercise for the caller.
+   */
+  cacheData: IResolverCacheFile;
+  /**
+   * The directory separator used in the `path` field of the resolver inputs.
+   * Will usually be `path.sep`.
+   */
+  resolverPathSeparator?: '/' | '\\';
+}
+
+function preservePath(path: string): string {
+  return path;
+}
+
+function backslashToSlash(path: string): string {
+  return path.replace(/\\/g, '/');
+}
+
+function slashToBackslash(path: string): string {
+  return path.replace(/\//g, '\\');
+}
+
+/**
+ * A cache of workspace layout information.
+ * @beta
+ */
+export class WorkspaceLayoutCache {
+  /**
+   * A lookup of context roots to their corresponding context objects
+   */
+  public readonly contextLookup: LookupByPath<IResolveContext>;
+  /**
+   * A weak map of package JSON contents to their corresponding context objects
+   */
+  public readonly contextForPackage: WeakMap<object, IResolveContext>;
+
+  public readonly resolverPathSeparator: string;
+  public readonly normalizeToSlash: (input: string) => string;
+  public readonly normalizeToPlatform: (input: string) => string;
+
+  public constructor(options: IWorkspaceLayoutCacheOptions) {
+    const { workspaceRoot, cacheData, resolverPathSeparator = directorySeparator } = options;
+
+    if (resolverPathSeparator !== '/' && resolverPathSeparator !== '\\') {
+      throw new Error(`Unsupported directory separator: ${resolverPathSeparator}`);
+    }
+
+    const resolveContexts: IResolveContext[] = [];
+    const contextLookup: LookupByPath<IResolveContext> = new LookupByPath(undefined, resolverPathSeparator);
+
+    this.contextLookup = contextLookup;
+    this.contextForPackage = new WeakMap<object, IResolveContext>();
+
+    const normalizeToSlash: (path: string) => string =
+      resolverPathSeparator === '/' ? preservePath : backslashToSlash;
+    const normalizeToPlatform: (path: string) => string =
+      resolverPathSeparator === '/' ? preservePath : slashToBackslash;
```

Comment:
Get rid of `preservePath` and instead use an `undefined` and call these with `normalizeToSlash?.(path) ?? path` to avoid a new stack frame in that case.

Reply:
That's a bit less ergonomic in the various callsites, since then I need to assign the input expression to a local variable to avoid duplicating work.

Reply:
Well, did it, but the result does feel a bit less readable.

Reply:
I'll need to do some comparative profiling once I have this up and running with real data to see which option is better. Since this is very much critical path code every bit of performance matters.